### DISCUSSION TITLE
I have my own architecture I'm working on, this is just the best place to ask. Issues with DDP.

### DISCUSSION THREAD

**Discussion by rlewkowicz** [ultralytics/ultralytics#21636] [Q&A]
I don't expect anyone to dig on my code. This based on yolov9 which is already ancient yolov5. It is no longer anything close to that code base or yolov5. It's truly a distinct entity. I'm coming for your yolov11 n model actually. In fact, I believe I will blow all current n models out of the water for size and latency. But stand by on that comment.

https://github.com/rlewkowicz/yolov9001

My question generally circles around DDP. I implemented refined lion, locally, works great. I matched their paper and it seems to be going in the right direction. DDP? Cant even get off the ground. 

Then, we flip it. AdamW, crushes DDP, locally elbows. 

I use sync batch norm, when doing ddp, for my compute loss instead of per gpu, I calculate global loss. My ema is properly placed etc.

So this is a general question. What are all things that are going to bite me on DDP? I expect some variance but It's wildly different and I'd like to pin it down. Ideally the configs will work with DDP and local.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rlewkowicz, thanks for starting the discussion and for your interest in Ultralytics 🚀 This is an automated response — an Ultralytics engineer will review your post and assist soon.

We recommend a visit to the Docs for quick references and examples that often address common questions:
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/
- Main Docs: https://docs.ultralytics.com/

If this is a 🐛 Bug Report, please include a minimum reproducible example (MRE) so we can help debug it efficiently:
- MRE Guide: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please share as much detail as possible:
- Your environment (OS, Python, PyTorch/CUDA, GPUs, DDP launch command)
- Exact configs/hyperparameters for local vs DDP runs
- Relevant logs (loss curves, LR schedules, batch sizes, SyncBN setup)
- Dataset samples and a few training logs
And please verify you’re following our Tips for Best Training Results:
- https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord for real-time chat: https://discord.com/invite/ultralytics 🎧
- Discourse for in-depth discussions: https://community.ultralytics.com/
- Subreddit for community threads: https://www.reddit.com/r/Ultralytics/

## Upgrade
Please ensure you’re on the latest ultralytics package and requirements in a Python>=3.8 environment with PyTorch>=1.8 to rule out resolved issues:
```bash
pip install -U ultralytics
```
Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments
YOLO can run in any of these verified environments (with CUDA/CUDNN, Python, and PyTorch preinstalled):
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us pinpoint DDP-specific issues, please share a minimal script that reproduces the divergence between local and DDP runs, along with the exact launch command and configs you’re using. This will help us provide targeted guidance quickly 🙌

**Reply by glenn-jocher** [reply]
Great question—and nice work pushing on your own arch. The biggest DDP gotchas we see are LR/global-batch mismatches (apply linear LR scaling or keep global batch constant), double-averaging the loss (compute per-rank mean and let DDP average gradients; only all-reduce for logging), BN not actually synced (convert with torch.nn.SyncBatchNorm.convert_sync_batchnorm before wrapping), sampler/seed drift (use DistributedSampler with consistent shuffle/drop_last and seed Python/NumPy/torch plus worker_init_fn), and side effects running on every rank (do dataset caching, EMA updates, and checkpoint writes on rank 0 only with a barrier). For reference, our trainer.train DDP launch logic shows Multi-GPU caveats like disabling rect and AutoBatch, torch_distributed_zero_first demonstrates how we serialize rank-0 setup, and the DDP utilities detail how we spawn and clean up distributed runs. Links: trainer.train DDP launch logic, torch_distributed_zero_first, and dist utilities.
